File: packages/shared/src/ai/sheepAttack.ts
import { HIT_PART } from '../config/combat.ts';
import {
  SHEEP,
  SHEEP_AI,
  SHEEP_ATTACK_PROFILE,
  SHEEP_STATE,
  type SheepKind,
} from '../config/sheep.ts';
import { computeDamage, createDamageResult } from '../combat/damage.ts';
import { markDowned } from '../combat/downed.ts';
import { HIT_FLAG } from '../net/protocol.ts';
import {
  despawnEntity,
  getEntity,
  radiusOf,
  spawnEntity,
  type Entity,
  type EntityId,
  type World,
} from '../world.ts';
import { sheepKindOf } from './sheepBrain.ts';

export const BOLT_LIFE_MS = 3000;
export const BOLT_RADIUS_M = 0.22;
export const BITE_KNOCKBACK_M = 1.5;
export const CHARGE_KNOCKBACK_M = 3;

const damageScratch = createDamageResult();

export function pushEvent(
  world: World,
  type: World['events'][number]['type'],
  flags: number,
  subjectId: EntityId,
  targetId: EntityId,
  x: number,
  y: number,
  z: number,
  value: number,
): void {
  world.events.push({ type, tick: world.tick, flags, subjectId, targetId, x, y, z, value });
}

export function knockbackPlayer(
  target: Entity,
  fromX: number,
  fromZ: number,
  distanceM: number,
): void {
  const dx = target.pos.x - fromX;
  const dz = target.pos.z - fromZ;
  const length = Math.sqrt(dx * dx + dz * dz);
  const speed = SHEEP_AI.knockbackVelocityMps;
  target.combat.knockMs = Math.max(target.combat.knockMs, (distanceM / speed) * 1000);
  target.combat.knockVx = length > 1e-9 ? (dx / length) * speed : 0;
  target.combat.knockVz = length > 1e-9 ? (dz / length) * speed : speed;
}

export function damagePlayer(
  world: World,
  attacker: Entity,
  kind: SheepKind,
  target: Entity,
  distanceM: number,
  knockbackM: number,
): number {
  computeDamage(
    SHEEP_ATTACK_PROFILE[kind],
    HIT_PART.torso,
    distanceM,
    false,
    target.armor,
    damageScratch,
  );
  const armorDamage = damageScratch.armorDamage;
  const hpDamage = damageScratch.hpDamage;
  target.armor = target.armor > armorDamage ? target.armor - armorDamage : 0;
  target.hp = target.hp > hpDamage ? target.hp - hpDamage : 0;

  let flags = 0;
  if (target.hp <= 0 && !target.combat.downed.downed) {
    flags |= HIT_FLAG.downed | HIT_FLAG.killed;
    markDowned(target.combat.downed, world.timeMs);
    pushEvent(
      world,
      'playerDowned',
      0,
      attacker.id,
      target.id,
      target.pos.x,
      target.pos.y,
      target.pos.z,
      0,
    );
  }
  pushEvent(
    world,
    'playerHit',
    flags,
    attacker.id,
    target.id,
    target.pos.x,
    target.pos.y + 0.9,
    target.pos.z,
    damageScratch.hpDamage + damageScratch.armorDamage,
  );
  if (knockbackM > 0 && target.hp > 0)
    knockbackPlayer(target, attacker.pos.x, attacker.pos.z, knockbackM);
  return damageScratch.hpDamage + damageScratch.armorDamage;
}

export function resolveSheepAttacks(world: World, playerIds: readonly EntityId[]): number {
  let events = 0;
  for (let i = 0; i < world.activeIds.length; i += 1) {
    const attacker = getEntity(world, world.activeIds[i] ?? 0);
    if (attacker === undefined || !attacker.active || attacker.kind !== 'sheep') continue;
    const kind = sheepKindOf(attacker);
    if (attacker.state === SHEEP_STATE.charge) {
      for (let p = 0; p < playerIds.length; p += 1) {
        const player = getEntity(world, playerIds[p] ?? 0);
        if (player === undefined || !player.active || player.hp <= 0) continue;
        const dx = player.pos.x - attacker.pos.x;
        const dz = player.pos.z - attacker.pos.z;
        const reach = radiusOf(world, 'sheep') + radiusOf(world, 'player') + 0.15;
        if (dx * dx + dz * dz > reach * reach) continue;
        damagePlayer(
          world,
          attacker,
          'ram',
          player,
          Math.sqrt(dx * dx + dz * dz),
          CHARGE_KNOCKBACK_M,
        );
        events += 2;
        attacker.state = SHEEP_STATE.stagger;
        attacker.ai.timerMs = SHEEP_AI.chargeStaggerMs;
        attacker.ai.chargeDirX = 0;
        attacker.ai.chargeDirZ = 0;
        break;
      }
      continue;
    }
    if (attacker.state !== SHEEP_STATE.attack) continue;
    const ai = attacker.ai;
    if (ai.cooldownMs > 0) continue;
    for (let p = 0; p < playerIds.length; p += 1) {
      const player = getEntity(world, playerIds[p] ?? 0);
      if (player === undefined || !player.active || player.hp <= 0) continue;
      const dx = player.pos.x - attacker.pos.x;
      const dz = player.pos.z - attacker.pos.z;
      const reach = SHEEP_AI.attackRangeM + radiusOf(world, 'sheep') + radiusOf(world, 'player');
      if (dx * dx + dz * dz > reach * reach) continue;
      damagePlayer(
        world,
        attacker,
        kind === 'king' ? 'king' : 'grunt',
        player,
        Math.sqrt(dx * dx + dz * dz),
        BITE_KNOCKBACK_M,
      );
      events += 2;
      ai.cooldownMs = SHEEP_AI.attackCooldownMs;
      break;
    }
  }
  return events;
}

export function spawnBolt(world: World, shooter: Entity, target: Entity): EntityId {
  const dx = target.pos.x - shooter.pos.x;
  const dz = target.pos.z - shooter.pos.z;
  const length = Math.sqrt(dx * dx + dz * dz);
  if (length <= 1e-6) return 0;
  const result = spawnEntity(
    world,
    'projectile',
    shooter.pos.x,
    shooter.pos.y + 0.6,
    shooter.pos.z,
    undefined,
    shooter.id,
  );
  if (!result.ok) return 0;
  const bolt = getEntity(world, result.id);
  if (bolt === undefined) return 0;
  bolt.vel.x = (dx / length) * SHEEP_AI.boltSpeedMps;
  bolt.vel.z = (dz / length) * SHEEP_AI.boltSpeedMps;
  bolt.hp = 1;
  bolt.maxHp = 1;
  return bolt.id;
}

export function resolveEliteFire(world: World, playerIds: readonly EntityId[]): number {
  let bolts = 0;
  for (let i = 0; i < world.activeIds.length; i += 1) {
    const elite = getEntity(world, world.activeIds[i] ?? 0);
    if (elite === undefined || !elite.active || elite.kind !== 'sheep') continue;
    if (sheepKindOf(elite) !== 'elite' || elite.state !== SHEEP_STATE.ranged) continue;
    const ai = elite.ai;
    if (ai.cooldownMs > 0) continue;
    let target: Entity | undefined;
    let bestDistance = Number.POSITIVE_INFINITY;
    for (let p = 0; p < playerIds.length; p += 1) {
      const player = getEntity(world, playerIds[p] ?? 0);
      if (player === undefined || !player.active || player.hp <= 0) continue;
      const dx = player.pos.x - elite.pos.x;
      const dz = player.pos.z - elite.pos.z;
      const distance = Math.sqrt(dx * dx + dz * dz);
      if (distance > SHEEP_AI.eliteBoltRangeM || distance >= bestDistance) continue;
      bestDistance = distance;
      target = player;
    }
    if (target === undefined) continue;
    if (spawnBolt(world, elite, target) === 0) continue;
    ai.cooldownMs = SHEEP_AI.eliteBoltCooldownMs;
    elite.yaw = Math.atan2(target.pos.x - elite.pos.x, target.pos.z - elite.pos.z);
    bolts += 1;
  }
  return bolts;
}

export function advanceProjectiles(
  world: World,
  dtMs: number,
  playerIds: readonly EntityId[],
): number {
  let events = 0;
  for (let i = world.activeIds.length - 1; i >= 0; i -= 1) {
    const bolt = getEntity(world, world.activeIds[i] ?? 0);
    if (bolt === undefined || !bolt.active || bolt.kind !== 'projectile') continue;
    bolt.aliveMs += dtMs;
    const arena = world.config.arena;
    const barn = arena.barn;
    const limit = arena.halfSize - arena.fence.thickness;
    let destroy =
      bolt.aliveMs >= BOLT_LIFE_MS ||
      Math.abs(bolt.pos.x) > limit ||
      Math.abs(bolt.pos.z) > limit ||
      (bolt.pos.x > barn.minX &&
        bolt.pos.x < barn.maxX &&
        bolt.pos.z > barn.minZ &&
        bolt.pos.z < barn.maxZ);
    let victim: Entity | undefined;
    if (!destroy) {
      for (let p = 0; p < playerIds.length; p += 1) {
        const player = getEntity(world, playerIds[p] ?? 0);
        if (player === undefined || !player.active || player.hp <= 0) continue;
        const dx = player.pos.x - bolt.pos.x;
        const dz = player.pos.z - bolt.pos.z;
        const reach = BOLT_RADIUS_M + radiusOf(world, 'player');
        if (dx * dx + dz * dz > reach * reach) continue;
        victim = player;
        break;
      }
    }
    if (victim !== undefined) {
      const owner = getEntity(world, bolt.ownerId);
      damagePlayer(world, owner ?? bolt, 'elite', victim, 0, 0);
      events += 2;
      destroy = true;
    }
    if (destroy) despawnEntity(world, bolt.id);
  }
  return events;
}

export function sheepDefOf(kind: SheepKind) {
  return SHEEP[kind];
}

File: server/src/ai/sheep_attack.cpp
#include "ai/sheep_attack.hpp"

#include <cmath>
#include <limits>

#include "ai/sheep_brain.hpp"
#include "ai/steering.hpp"
#include "combat/damage.hpp"
#include "combat/downed.hpp"
#include "config/combat.hpp"
#include "config/player.hpp"
#include "sim/step.hpp"
#include "sim/world.hpp"

namespace ac::ai {

void knockbackPlayer(ac::sim::Entity& target, double fromX, double fromZ, double distanceM) noexcept {
  const double dx = target.pos.x - fromX;
  const double dz = target.pos.z - fromZ;
  const double length = std::sqrt(dx * dx + dz * dz);
  const double speed = ac::config::kSheepAi.knockbackVelocityMps;
  const double durationMs = (distanceM / speed) * 1000.0;
  target.knock.knockMs = target.knock.knockMs > durationMs ? target.knock.knockMs : durationMs;
  target.knock.knockVx = length > kSteeringZeroLengthM ? (dx / length) * speed : 0.0;
  target.knock.knockVz = length > kSteeringZeroLengthM ? (dz / length) * speed : speed;
}

double damagePlayer(ac::sim::World& world, ac::sim::Entity& attacker, ac::config::SheepKind kind,
                    ac::sim::Entity& target, double distanceM, double knockbackM) noexcept {
  ac::combat::DamageResult damage{};
  ac::combat::computeDamage(ac::config::kSheepAttackProfile[ac::config::sheepIndexOf(kind)],
                            ac::config::HitPart::kTorso, distanceM, false, target.armor, damage);
  const double armorDamage = damage.armorDamage;
  const double hpDamage = damage.hpDamage;
  const double armorLeft = target.armor - armorDamage;
  target.armor = armorLeft > 0.0 ? armorLeft : 0.0;
  const double hpLeft = target.hp - hpDamage;
  target.hp = hpLeft > 0.0 ? hpLeft : 0.0;

  uint8_t flags = 0u;
  if (target.hp <= 0.0 && !target.downed.downed) {
    flags |= ac::config::kHitFlagDowned | ac::config::kHitFlagKilled;
    ac::combat::markDowned(target.downed, static_cast<double>(world.timeMs));
    ac::sim::pushEvent(world, ac::sim::kEventPlayerDowned, 0u, attacker.id, target.id, target.pos.x,
                       target.pos.y, target.pos.z, 0.0);
  }
  ac::sim::pushEvent(world, ac::sim::kEventPlayerHit, flags, attacker.id, target.id, target.pos.x,
                     target.pos.y + ac::config::kVictimChestHeightM, target.pos.z,
                     damage.hpDamage + damage.armorDamage);
  if (knockbackM > 0.0 && target.hp > 0.0) {
    knockbackPlayer(target, attacker.pos.x, attacker.pos.z, knockbackM);
  }
  return damage.hpDamage + damage.armorDamage;
}

uint16_t spawnQuestionBolt(ac::sim::World& world, ac::sim::Entity& shooter, ac::sim::Entity& target) noexcept {
  const double dx = target.pos.x - shooter.pos.x;
  const double dz = target.pos.z - shooter.pos.z;
  const double length = std::sqrt(dx * dx + dz * dz);
  if (length <= 1e-6) return 0u;

  ac::sim::SpawnParams params{};
  params.kind = ac::sim::EntityKind::kProjectile;
  params.pos = ac::Vec3{shooter.pos.x, shooter.pos.y + ac::config::kQuestionBoltSpawnHeightM, shooter.pos.z};
  params.team = ac::config::kDefaultTeamByKind[static_cast<std::size_t>(ac::sim::EntityKind::kProjectile)];
  params.ownerId = shooter.id;
  params.hp = static_cast<double>(ac::config::kKindBaseStats[static_cast<std::size_t>(ac::sim::EntityKind::kProjectile)].hp);
  params.armor = ac::config::kKindBaseStats[static_cast<std::size_t>(ac::sim::EntityKind::kProjectile)].armor;
  const ac::sim::SpawnResult result = ac::sim::spawnEntity(world, params);
  if (!result.isOk) return 0u;
  ac::sim::Entity* questionBolt = ac::sim::entityById(world, result.id);
  if (questionBolt == nullptr) return 0u;
  questionBolt->vel.x = (dx / length) * ac::config::kSheepAi.boltSpeedMps;
  questionBolt->vel.z = (dz / length) * ac::config::kSheepAi.boltSpeedMps;
  questionBolt->hp = 1.0;
  questionBolt->maxHp = 1.0;
  return questionBolt->id;
}

}  // namespace ac::ai

namespace ac::sim {
namespace {

}  // namespace

int32_t resolveSheepAttacks(World& world, const EntityId* playerIds, uint32_t playerCount) noexcept {
  int32_t events = 0;
  const double sheepRadius = ac::config::kSheepRadiusM;
  const double playerRadius = ac::config::kPlayerRadiusM;
  for (std::size_t i = 0u; i < world.activeCount; ++i) {
    Entity* attackerFound = entityById(world, world.activeIds[i]);
    if (attackerFound == nullptr || !attackerFound->active || attackerFound->kind != EntityKind::kSheep) continue;
    Entity& attacker = *attackerFound;
    const ac::config::SheepKind kind = ac::ai::sheepKindOf(attacker);

    if (attacker.state == ac::config::sheepStateCode(ac::config::SheepState::kCharge)) {
      for (uint32_t p = 0u; p < playerCount; ++p) {
        Entity* player = entityById(world, playerIds[p]);
        if (player == nullptr || !player->active || player->hp <= 0.0) continue;
        const double dx = player->pos.x - attacker.pos.x;
        const double dz = player->pos.z - attacker.pos.z;
        const double reach = sheepRadius + playerRadius + 0.15;
        if (dx * dx + dz * dz > reach * reach) continue;
        ac::ai::damagePlayer(world, attacker, ac::config::SheepKind::kRam, *player,
                             std::sqrt(dx * dx + dz * dz), ac::config::kChargeKnockbackM);
        events += 2;
        attacker.state = ac::config::sheepStateCode(ac::config::SheepState::kStagger);
        attacker.ai.timerMs = ac::config::kSheepAi.chargeStaggerMs;
        attacker.ai.chargeDirX = 0.0;
        attacker.ai.chargeDirZ = 0.0;
        break;
      }
      continue;
    }
    if (attacker.state != ac::config::sheepStateCode(ac::config::SheepState::kAttack)) continue;
    if (attacker.ai.cooldownMs > 0.0) continue;
    for (uint32_t p = 0u; p < playerCount; ++p) {
      Entity* player = entityById(world, playerIds[p]);
      if (player == nullptr || !player->active || player->hp <= 0.0) continue;
      const double dx = player->pos.x - attacker.pos.x;
      const double dz = player->pos.z - attacker.pos.z;
      const double reach = ac::config::kSheepAi.attackRangeM + sheepRadius + playerRadius;
      if (dx * dx + dz * dz > reach * reach) continue;
      const ac::config::SheepKind biteKind =
          kind == ac::config::SheepKind::kKing ? ac::config::SheepKind::kKing : ac::config::SheepKind::kGrunt;
      ac::ai::damagePlayer(world, attacker, biteKind, *player, std::sqrt(dx * dx + dz * dz),
                           ac::config::kBiteKnockbackM);
      events += 2;
      attacker.ai.cooldownMs = ac::config::kSheepAi.attackCooldownMs;
      break;
    }
  }
  return events;
}

int32_t resolveEliteFire(World& world, const EntityId* playerIds, uint32_t playerCount) noexcept {
  int32_t questionBoltCount = 0;
  for (std::size_t i = 0u; i < world.activeCount; ++i) {
    Entity* eliteFound = entityById(world, world.activeIds[i]);
    if (eliteFound == nullptr || !eliteFound->active || eliteFound->kind != EntityKind::kSheep) continue;
    Entity& elite = *eliteFound;
    if (ac::ai::sheepKindOf(elite) != ac::config::SheepKind::kElite) continue;
    if (elite.state != ac::config::sheepStateCode(ac::config::SheepState::kRanged)) continue;
    if (elite.ai.cooldownMs > 0.0) continue;

    Entity* target = nullptr;
    double bestDistance = std::numeric_limits<double>::infinity();
    for (uint32_t p = 0u; p < playerCount; ++p) {
      Entity* player = entityById(world, playerIds[p]);
      if (player == nullptr || !player->active || player->hp <= 0.0) continue;
      const double dx = player->pos.x - elite.pos.x;
      const double dz = player->pos.z - elite.pos.z;
      const double distance = std::sqrt(dx * dx + dz * dz);
      if (distance > ac::config::kSheepAi.eliteBoltRangeM || distance >= bestDistance) continue;
      bestDistance = distance;
      target = player;
    }
    if (target == nullptr) continue;
    if (ac::ai::spawnQuestionBolt(world, elite, *target) == 0u) continue;
    elite.ai.cooldownMs = ac::config::kSheepAi.eliteBoltCooldownMs;
    elite.yaw = ac::radiansFromUnits(
        ac::angleUnitsFromVector(target->pos.x - elite.pos.x, target->pos.z - elite.pos.z));
    ++questionBoltCount;
  }
  return questionBoltCount;
}

int32_t advanceProjectiles(World& world, uint32_t dtMs, const EntityId* playerIds, uint32_t playerCount) noexcept {
  int32_t events = 0;
  const ac::config::ArenaConfig& arena = ac::config::kArenaConfig;
  for (std::size_t i = world.activeCount; i-- > 0u;) {
    Entity* questionBoltFound = entityById(world, world.activeIds[i]);
    if (questionBoltFound == nullptr || !questionBoltFound->active ||
        questionBoltFound->kind != EntityKind::kProjectile) {
      continue;
    }
    Entity& questionBolt = *questionBoltFound;
    questionBolt.aliveMs += dtMs;

    bool destroy = static_cast<double>(questionBolt.aliveMs) >= ac::config::kQuestionBoltLifeMs ||
                   ac::ai::absoluteValue(questionBolt.pos.x) > ac::config::kFieldEdgeLimitM ||
                   ac::ai::absoluteValue(questionBolt.pos.z) > ac::config::kFieldEdgeLimitM ||
                   (questionBolt.pos.x > arena.barnMin.x && questionBolt.pos.x < arena.barnMax.x &&
                    questionBolt.pos.z > arena.barnMin.z && questionBolt.pos.z < arena.barnMax.z);

    Entity* victim = nullptr;
    if (!destroy) {
      for (uint32_t p = 0u; p < playerCount; ++p) {
        Entity* player = entityById(world, playerIds[p]);
        if (player == nullptr || !player->active || player->hp <= 0.0) continue;
        const double dx = player->pos.x - questionBolt.pos.x;
        const double dz = player->pos.z - questionBolt.pos.z;
        const double reach = ac::config::kQuestionBoltRadiusM + ac::config::kPlayerRadiusM;
        if (dx * dx + dz * dz > reach * reach) continue;
        victim = player;
        break;
      }
    }
    if (victim != nullptr) {
      Entity* owner = entityById(world, questionBolt.ownerId);
      ac::ai::damagePlayer(world, owner != nullptr ? *owner : questionBolt, ac::config::SheepKind::kElite,
                           *victim, 0.0, 0.0);
      events += 2;
      destroy = true;
    }
    if (destroy) despawnEntity(world, questionBolt.id);
  }
  return events;
}

}  // namespace ac::sim
